File: 189轮转数组/rotate_array.py
from __future__ import annotations


def rotate(nums: list[int], k: int) -> None:
    n = len(nums)
    if n <= 1:
        return
    k %= n
    if k == 0:
        return

    def rev(l: int, r: int) -> None:
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1

    rev(0, n - 1)
    rev(0, k - 1)
    rev(k, n - 1)


def _parse_ints(line: str) -> list[int]:
    # allow formats like:
    # "1 2 3", "[1,2,3]", "nums = [1, 2, 3]"
    cleaned = (
        line.replace("[", " ")
        .replace("]", " ")
        .replace(",", " ")
        .replace("nums", " ")
        .replace("=", " ")
    )
    out: list[int] = []
    for tok in cleaned.split():
        try:
            out.append(int(tok))
        except ValueError:
            continue
    return out


def _format_list(nums: list[int]) -> str:
    return "[" + ",".join(str(x) for x in nums) + "]"


def main() -> None:
    """
    Interactive stdin mode (until EOF):
    - line1: nums (space-separated, or [..], or "nums = [..]")
    - line2: k
    Output each case as: [rotated_array]
    """
    import sys

    it = iter(sys.stdin.readline, "")
    for line_nums in it:
        line_nums = line_nums.strip()
        if not line_nums:
            continue
        nums = _parse_ints(line_nums)

        line_k = next(it, "")
        if not line_k:
            break
        k_tokens = _parse_ints(line_k)
        if not k_tokens:
            continue
        k = k_tokens[0]

        rotate(nums, k)
        sys.stdout.write(_format_list(nums) + "\n")


if __name__ == "__main__":
    main()

File: 189轮转数组/rotate_array.cpp
#include <algorithm>
#include <cctype>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

static std::vector<long long> parseInts(const std::string& line) {
    // allow: "1 2 3", "[1,2,3]", "nums = [1, 2, 3]"
    std::string s = line;
    for (char& c : s) {
        if (std::isdigit((unsigned char)c) || c == '-') continue;
        c = ' ';
    }
    std::stringstream ss(s);
    std::vector<long long> out;
    long long x;
    while (ss >> x) out.push_back(x);
    return out;
}

static void rotate(std::vector<long long>& nums, long long k) {
    int n = (int)nums.size();
    if (n <= 1) return;
    k %= n;
    if (k == 0) return;
    std::reverse(nums.begin(), nums.end());
    std::reverse(nums.begin(), nums.begin() + k);
    std::reverse(nums.begin() + k, nums.end());
}

static void printList(const std::vector<long long>& nums) {
    std::cout << "[";
    for (size_t i = 0; i < nums.size(); ++i) {
        if (i) std::cout << ",";
        std::cout << nums[i];
    }
    std::cout << "]\n";
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    // Interactive stdin mode (until EOF):
    // - line1: nums (space-separated, or [..], or "nums = [..]")
    // - line2: k
    // Output each case as: [rotated_array]
    std::string line_nums;
    while (std::getline(std::cin, line_nums)) {
        if (line_nums.find_first_not_of(" \t\r\n") == std::string::npos) continue;
        auto nums_ll = parseInts(line_nums);
        std::vector<long long> nums(nums_ll.begin(), nums_ll.end());

        std::string line_k;
        if (!std::getline(std::cin, line_k)) break;
        auto ks = parseInts(line_k);
        if (ks.empty()) continue;
        long long k = ks[0];

        rotate(nums, k);
        printList(nums);
    }
    return 0;
}


